teps/0125-configmaps-and-secrets-as-param-value-source.md

```diff
@@ -0,0 +1,175 @@
+- Minimises changing of `Pipeline` or `Task` based on source of the value when executing in multiple clusters / namespace promoting reusability.
+
+
+## Requirements
+
+- Users should be able to pass `ConfigMap` or `Secret` reference as part of `TaskRun` or `PipelineRun` params along with value.
+
+## Proposal
+
+To achieve the requirement, lets take inspiration from `Kubernetes API Env section of each Container`
+
+```go
+// EnvVar represents an environment variable present in a Container.
+
+type  EnvVar  struct {
+  Name string  `json:"name" protobuf:"bytes,1,opt,name=name"`
+  Value string  `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+  ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+type  EnvVarSource  struct {
+  FieldRef *ObjectFieldSelector `json:"fieldRef,omitempty" protobuf:"bytes,1,opt,name=fieldRef"`
+  ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty" protobuf:"bytes,2,opt,name=resourceFieldRef"`
+  ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+  SecretKeyRef *SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+
+```
+
+From context of `PipelineRun` and `TaskRun` params 
+```yaml
+- FieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference field of the pod.
+- ResourceFieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference resource of the container.
+- ConfigMapKeyRef : To reference ConfigMap value
+- SecretKeyRef : To reference Secret value
+```
+
+
+### API
+
+To support the feature supported outlined above, we also propose a small addition to the API. 
+In `ParamSpec` need to `TaskRef` and `PipelineRef` objects will include a `ValueFrom` field.
+
+```go
+type  Param  struct {
+  Name string  `json:"name"`
+  Value ArrayOrString `json:"value"`
+  // Additional field ValueFrom to fetch value from ConfigMap or Secret
+  ValueFrom *ValueSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// ValueSource represents a source for the value.
+type  ValueSource  struct {
+  ConfigMapKeyRef *corev1.ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+  SecretKeyRef *corev1.SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+```
+
+For users trying to refer param in  `PipelineRun` or `TaskRun` the options would look like
+
+```yaml
+spec:
+  params:
+    - name: FROM_CONFIGMAP
+      valueFrom:
+       configMapKeyRef:
+        name: pipeline-run-configmap
+        key: HELLO_WORLD_KEY
+    - name: FROM_SECRET
+      valueFrom:
+       secretKeyRef:
+        name: pipeline-run-secret
+        key: HELLO_WORLD_KEY
```

Comment:
param values are stored in cleartext as part of the CRD. Maybe I'm missing something but does this include plans for scrubbing param values that come from secrets?

Reply:
That's a good point @lbernick. If a `PipelineRun` parameter is passed down to a `TaskRun`, it could use the same `valueFrom` syntax, but only when passed alone (i.e. not concatenated to other strings). The `TaskRun` controller again could use the same syntax for pods, but that would mean that the secret value could only be available to steps as an environment variable. Using secrets with environment variables can be already done today thanks to https://github.com/tektoncd/community/blob/main/teps/0101-env-in-pod-template.md